sklearn/ensemble/_hist_gradient_boosting/_predictor.pyx

```diff
@@ -17,50 +17,71 @@ from .common cimport Y_DTYPE_C
 from .common import Y_DTYPE
 from .common cimport X_BINNED_DTYPE_C
 from .common cimport node_struct
+from ._bitset cimport in_bitset
+from ._predictor_bitset cimport PredictorBitSet
+from ._cat_mapper cimport CategoryMapper
 
 np.import_array()
 
 
-def _predict_from_numeric_data(
+def _predict_from_data(
         node_struct [:] nodes,
+        PredictorBitSet predictor_bitset,
         const X_DTYPE_C [:, :] numeric_data,
         Y_DTYPE_C [:] out):
 
     cdef:
         int i
 
     for i in prange(numeric_data.shape[0], schedule='static', nogil=True):
-        out[i] = _predict_one_from_numeric_data(nodes, numeric_data, i)
+        out[i] = _predict_one_from_numeric_data(
+            nodes, predictor_bitset, numeric_data, i)
 
 
 cdef inline Y_DTYPE_C _predict_one_from_numeric_data(
         node_struct [:] nodes,
+        PredictorBitSet predictor_bitset,
         const X_DTYPE_C [:, :] numeric_data,
         const int row) nogil:
     # Need to pass the whole array and the row index, else prange won't work.
     # See issue Cython #2798
 
     cdef:
         node_struct node = nodes[0]
+        unsigned int node_idx = 0
 
     while True:
         if node.is_leaf:
             return node.value
 
         if isnan(numeric_data[row, node.feature_idx]):
             if node.missing_go_to_left:
-                node = nodes[node.left]
+                node_idx = node.left
             else:
-                node = nodes[node.right]
+                node_idx = node.right
+        elif node.is_categorical:
+            if not predictor_bitset.is_known_category(
+                    node.feature_idx, numeric_data[row, node.feature_idx]):
+                if node.missing_go_to_left:
+                    node_idx = node.left
+                else:
+                    node_idx = node.right
```

Comment:
make this a one-liner?
Also I'd just add a small comment:  # treat unknown categories as missing.